src/Catalyst.Core.Modules.Dfs/UnixFs/SizeChunker.cs

```diff
@@ -0,0 +1,199 @@
+                    }
+
+                    length += n;
+                    totalBytes += (uint) n;
+                }
+
+                //  Only generate empty block, when the stream is empty.
+                if (length == 0 && nodes.Count > 0)
+                {
+                    chunking = false;
+                    break;
+                }
+
+                if (options.Progress != null)
+                {
+                    options.Progress.Report(new TransferProgress
+                    {
+                        Name = name,
+                        Bytes = totalBytes
+                    });
+                }
+
+                // if protected data, then get CMS structure.
+                if (protecting)
+                {
+                    // TODO: Inefficent to copy chunk, use ArraySegment in DataMessage.Data
+                    var plain = new byte[length];
+                    Array.Copy(chunk, plain, length);
+                    var cipher = await keyChain.CreateProtectedDataAsync(options.ProtectionKey, plain, cancel)
+                       .ConfigureAwait(false);
+                    var cid = await blockService.PutAsync(
+                        cipher,
+                        "cms",
+                        options.Hash,
+                        options.Encoding,
+                        options.Pin,
+                        cancel).ConfigureAwait(false);
+                    nodes.Add(new UnixFsNode
+                    {
+                        Id = cid,
+                        Size = length,
+                        DagSize = cipher.Length,
+                        Links = UnixFsLink.None
+                    });
+                }
+                else if (options.RawLeaves)
+                {
+                    // TODO: Inefficent to copy chunk, use ArraySegment in DataMessage.Data
+                    var data = new byte[length];
+                    Array.Copy(chunk, data, length);
+                    var cid = await blockService.PutAsync(
+                        data,
+                        "raw",
+                        options.Hash,
+                        options.Encoding,
+                        options.Pin,
+                        cancel).ConfigureAwait(false);
+                    nodes.Add(new UnixFsNode
+                    {
+                        Id = cid,
+                        Size = length,
+                        DagSize = length,
+                        Links = UnixFsLink.None
+                    });
+                }
+                else
+                {
+                    // Build the DAG.
+                    var dm = new DataMessage
+                    {
+                        Type = DataType.File,
+                        FileSize = (ulong) length
+                    };
+                    if (length > 0)
+                    {
+                        // TODO: Inefficent to copy chunk, use ArraySegment in DataMessage.Data
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)